pkg/kv/dynamodb/store.go

```diff
@@ -0,0 +1,320 @@
+		Value:     []byte(item.ItemValue),
+		Predicate: kv.Predicate([]byte(item.ItemValue)),
+	}, nil
+}
+
+func (s *Store) Set(ctx context.Context, key, value []byte) error {
+	return s.setWithOptionalPredicate(ctx, key, value, nil, false)
+}
+
+func (s *Store) SetIf(ctx context.Context, key, value []byte, valuePredicate kv.Predicate) error {
+	return s.setWithOptionalPredicate(ctx, key, value, valuePredicate, true)
+}
+
+func (s *Store) setWithOptionalPredicate(ctx context.Context, key, value []byte, valuePredicate kv.Predicate, usePredicate bool) error {
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+
+	item := DynKVItem{
+		PartKey:   s.getPartitionKey(key),
+		ItemKey:   string(key),
+		ItemValue: string(value),
+	}
+
+	marshaledItem, err := dynamodbattribute.MarshalMap(item)
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+
+	input := &dynamodb.PutItemInput{
+		Item:      marshaledItem,
+		TableName: &s.params.TableName,
+	}
+	if usePredicate {
+		if valuePredicate != nil {
+			input.ConditionExpression = aws.String(ItemValue + " = :predicate")
+			input.ExpressionAttributeValues = map[string]*dynamodb.AttributeValue{
+				":predicate": {S: aws.String(string(valuePredicate.([]byte)))},
+			}
+		} else {
+			input.ConditionExpression = aws.String("attribute_not_exists(" + ItemValue + ")")
+		}
+	}
+
+	_, err = s.svc.PutItemWithContext(ctx, input)
+
+	if err != nil {
+		if _, ok := err.(*dynamodb.ConditionalCheckFailedException); ok && usePredicate {
+			return kv.ErrPredicateFailed
+		}
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+	return nil
+}
+
+func (s *Store) Delete(ctx context.Context, key []byte) error {
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+
+	_, err := s.svc.DeleteItemWithContext(ctx, &dynamodb.DeleteItemInput{
+		TableName: aws.String(s.params.TableName),
+		Key:       s.bytesKeyToDynamoKey(key),
+	})
+
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+	return nil
+}
+
+func (s *Store) Scan(ctx context.Context, start []byte) (kv.EntriesIterator, error) {
+	return s.scanInternal(ctx, start, nil)
+}
+
+func (s *Store) scanInternal(ctx context.Context, scanKey []byte, exclusiveStartKey map[string]*dynamodb.AttributeValue) (*EntriesIterator, error) {
+	keyConditionExpression := PartKey + " = :partkey"
+	expressionAttributeValues := map[string]*dynamodb.AttributeValue{
+		":partkey": {
+			S: aws.String(s.getPartitionKey(scanKey)),
+		},
+	}
+	if len(scanKey) > 0 {
+		keyConditionExpression += " AND " + ItemKey + " >= :fromkey"
+		expressionAttributeValues[":fromkey"] = &dynamodb.AttributeValue{
+			S: aws.String(string(scanKey)),
+		}
+	}
+	queryInput := &dynamodb.QueryInput{
+		TableName:                 aws.String(s.params.TableName),
+		KeyConditionExpression:    aws.String(keyConditionExpression),
+		ExpressionAttributeValues: expressionAttributeValues,
```

Comment:
Why are both of them needed?

Reply:
`KeyConditionExpression` is the scan term, `PartitionaLey = :partitionkey AND ItemKey >= :fromkey` where `:partitionkey` and `:fromkey` are attributes values, specified in `ExpressionAttributeValues`, in json format following dynamodb definitions.
I'm not sure if it is possible to skip the usage of `ExpressionAttributeValues` and specify the values in the `KeyConditionExpression` itself but I spend sometime on it and couldn't get it to work (at least not for `awscli`).
This format also aligns with all the examples from `aws`